src/compas/datastructures/assembly/assembly.py

```diff
@@ -0,0 +1,142 @@
+        super(Assembly, None).__init__()
+        self.attributes = {}
+        self.graph = Graph()
+        self.parts = {}
+
+    def __str__(self):
+        tpl = "<Assembly with {} parts and {} connections>"
+        return tpl.format(self.graph.number_of_nodes(), self.graph.number_of_edges())
+
+    @property
+    def data(self):
+        """dict : A data dict representing the assembly data structure for serialization.
+        """
+        data = {
+            'attributes': self.attributes,
+            'graph': self.graph.data,
+        }
+        return data
+
+    @data.setter
+    def data(self, data):
+        self.attributes.update(data['attributes'] or {})
+        self.graph.data = data['graph']
+
+    def add_part(self, part, key=None, **kwargs):
+        """Add a part to the assembly.
+
+        Parameters
+        ----------
+        part: :class:`compas.datastructures.Part`
+            The part to add.
+        key: int or str, optional
+            The identifier of the part in the assembly.
+            Note that the key is unique only in the context of the current assembly.
+            Nested assemblies may have the same ``key`` value for one of their parts.
+            Default is ``None`` in which case the key will be automatically assigned integer value.
+        kwargs: dict
+            Additional named parameters collected in a dict.
+
+        Returns
+        -------
+        int or str
+            The identifier of the part in the current assembly graph.
+
+        """
+        key = self.graph.add_node(key=key, part=part, **kwargs)
+        part.key = key
+        self.parts[part.guid] = part
+        return key
+
+    def add_connection(self, a, b, **kwargs):
+        """Add a connection between two parts.
+
+        Parameters
+        ----------
+        a: :class:`compas.datastructures.Part`
+            The "from" part.
+        b: :class:`compas.datastructures.Part`
+            The "to" part.
+        kwargs: dict
+            Additional named parameters collected in a dict.
+
+        Returns
+        -------
+        tuple of str or int
+            The tuple of node identifiers that identifies the connection.
+
+        Raises
+        ------
+        :class:`AssemblyError`
+            If ``a`` and/or ``b`` are not in the assembly.
+        """
+        if a.key is None or b.key is None:
+            raise AssemblyError('Both parts have to be added to the assembly before a connection can be created.')
+        if not self.graph.has_node(a.key) or not self.graph.has_node(b.key):
+            raise AssemblyError('Both parts have to be added to the assembly before a connection can be created.')
+        return self.graph.add_edge(a.key, b.key, **kwargs)
+
+    def parts(self):
+        """The parts of the assembly.
+
+        Yields
+        ------
+        :class:`compas.datastructures.Part`
+            The individual parts of the assembly.
+        """
+        for node in self.graph.nodes():
```

Comment:
this iteration is not stable across the cpython-ironpython divide. The iter order of cpython is not a match to the one in ironpython. Once option would be to explicitly do `sorted(self.graph.nodes())` to make sure they match.

Reply:
we could use an ordered dict explicitly to be compatible with ipy, but we would have to make the change on the graph directly ...

Reply:
Changing graph sounds to radical if only for this reason...

In the "fab domain", the concept of an assembly sequencer is important... perhaps one option is to implement the interface of such a class at the compas core level, add a basic implementation that does linear sequencing based on key sorting, and then this method can use that.